consensus/src/experimental/buffer_manager.rs

```diff
@@ -500,61 +377,50 @@ impl StateManager {
             }
             cursor = next_cursor;
         }
-        Ok(())
     }
 
     async fn start(mut self) {
         info!("Buffer manager starts.");
         let mut interval =
             tokio::time::interval(Duration::from_millis(BUFFER_MANAGER_RETRY_INTERVAL));
         while !self.end_epoch {
-            // process new messages
-            if let Err(e) = tokio::select! {
+            // advancing the root will trigger sending requests to the pipeline
+            tokio::select! {
                 Some(blocks) = self.block_rx.next() => {
-                    self.process_ordered_blocks(blocks).await
+                    self.process_ordered_blocks(blocks);
+                    if self.execution_root.is_none() {
+                        self.advance_execution_root().await;
+                    }
                 }
                 Some(reset_event) = self.sync_rx.next() => {
-                    self.process_sync_request(reset_event).await
+                    self.process_sync_request(reset_event).await;
+                    if self.execution_root.is_none() {
+                        self.advance_execution_root().await;
+                    }
+                    if self.signing_root.is_none() {
+                        self.advance_signing_root().await;
+                    }
                 }
-                Some(execution_resp) = self.execution_phase_rx.next() => {
-                    self.process_execution_response(execution_resp).await
+                Some(response) = self.execution_phase_rx.next() => {
+                    self.process_execution_response(response).await;
+                    self.advance_execution_root().await;
+                    if self.signing_root.is_none() {
+                        self.advance_signing_root().await;
+                    }
                 }
-                Some(signing_resp) = self.signing_phase_rx.next() => {
-                    self.process_signing_response(signing_resp).await
+                Some(response) = self.signing_phase_rx.next() => {
+                    self.process_signing_response(response).await;
+                    self.advance_signing_root().await;
                 }
                 Some(commit_msg) = self.commit_msg_rx.next() => {
-                    self.process_commit_msg(commit_msg).await
+                    self.process_commit_message(commit_msg).await;
+                    self.try_advance_head().await;
```

Comment:
similarly, advance the head only if the commit message matched an item?

Reply:
don't have strong opinion on this, we can return something if commit message aggregated an item and use advance_head